lib/byron/bench/Restore.hs

```diff
@@ -0,0 +1,468 @@
+
+initBenchmarkLogging :: Severity -> IO (CM.Configuration, Trace IO Text)
+initBenchmarkLogging minSeverity = do
+    c <- defaultConfigStdout
+    CM.setMinSeverity c minSeverity
+    CM.setSetupBackends c [CM.KatipBK, CM.AggregationBK]
+    (tr, _sb) <- setupTrace_ c "bench-restore"
+    pure (c, tr)
+
+{-------------------------------------------------------------------------------
+                                  Benchmarks
+-------------------------------------------------------------------------------}
+
+{-# ANN bench_restoration ("HLint: ignore Use camelCase" :: String) #-}
+bench_restoration
+    :: forall (n :: NetworkDiscriminant) (k :: Depth -> * -> *) s t.
+        ( IsOurs s Address
+        , IsOurs s ChimericAccount
+        , IsOwned s ShelleyKey
+        , WalletKey k
+        , NFData s
+        , Show s
+        , MaxSizeOf Address n k
+        , PersistState s
+        , PersistPrivateKey (k 'RootK)
+        , KnownNetwork n
+        , t ~ IO Byron
+        )
+    => Trace IO Text
+    -> FilePath
+       -- ^ Socket path
+    -> FilePath
+       -- ^ Log output
+    -> (WalletId, WalletName, s)
+    -> IO ()
+bench_restoration tracer socketPath progressLogFile (wid, wname, s) = do
+    let bp = blockchainParameters @n
+    let vData = versionData @n
+    let network = "mainnet" :: String
+    let pm = fromNetworkMagic $ networkMagic $ fst vData
+    let tl = newTransactionLayer @n @k @(IO Byron) (Proxy) pm
+    withNetworkLayer nullTracer bp socketPath vData $ \nw' -> do
+        let convert = fromByronBlock (getGenesisBlockHash bp) (getEpochLength bp)
+        let nw = convert <$> nw'
+        withBenchDBLayer @s @k tracer $ \db -> do
+            BlockHeader sl _ _ _ <- unsafeRunExceptT $ currentNodeTip nw
+            sayErr . fmt $ network ||+ " tip is at " +|| sl ||+ ""
+
+            withFile progressLogFile WriteMode $ \h -> do
+                -- Use a custom tracer to output (time, blockHeight) to a file
+                -- each time we apply blocks.
+                let fileTr = Tracer $ \msg -> do
+                        liftIO . B8.hPut h . T.encodeUtf8 . (<> "\n") $ msg
+                        hFlush h
+                let w = WalletLayer
+                        (traceProgressForPlotting fileTr)
+                        (emptyGenesis bp, bp, mkSyncTolerance 3600)
+                        nw
+                        tl
+                        db
+                wallet <- unsafeRunExceptT $ W.createWallet w wid wname s
+                void $ forkIO $ unsafeRunExceptT $ W.restoreWallet @_ @s @t @k w wid
+                waitForWalletSync @n w wallet
+                (wallet', _, pending) <- unsafeRunExceptT $ W.readWallet w wid
+                sayErr "Wallet restored!"
+                sayErr . fmt $ "Balance: " +|| totalBalance pending wallet' ||+ " lovelace"
+                sayErr . fmt $
+                    "UTxO: " +|| Map.size (getUTxO $ totalUTxO pending wallet') ||+ " entries"
```

Comment:
What about printing the UTxO statistics also here? To get an overview of the distribution.